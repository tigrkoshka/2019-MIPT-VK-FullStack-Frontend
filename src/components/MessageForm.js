import React from 'react'
import PropTypes from 'prop-types'
import Header from './Header'
import toChats from '../images/toChats.png'
import attach from '../images/attachment.png'
import geo from '../images/geolocation.png'
import voice from '../images/voice.png'
import stop from '../images/stopRecording.png'
import messageStyles from '../styles/singleMessageStyles.module.scss'
import formStyles from '../styles/messageFormStyles.module.scss'

function singleTextMessage({ content, time, whose, key }) {
  return (
    <div className={messageStyles.mine} style={{ maxWidth: '75%' }} key={key}>
      <div className={messageStyles.content}>{content}</div>
      <div className={messageStyles.time}>{time}</div>
    </div>
  )
}

singleTextMessage.propTypes = {
  content: PropTypes.string.isRequired,
  time: PropTypes.string.isRequired,
  whose: PropTypes.string.isRequired,
  key: PropTypes.number.isRequired,
}

function singleImageMessage({ url, time, whose, key }) {
  return (
    <div className={messageStyles.mine} style={{ background: 'rgb(40, 40, 40)' }} key={key}>
      <img src={url} className={messageStyles.image} alt="blob was deleted" />
      <div className={messageStyles.time}>{time}</div>
    </div>
  )
}

singleImageMessage.propTypes = {
  url: PropTypes.string.isRequired,
  time: PropTypes.string.isRequired,
  whose: PropTypes.string.isRequired,
  key: PropTypes.number.isRequired,
}

function singleAudioMessage({ url, time, whose, key }) {
  return (
    <div className={messageStyles.mine} style={{ width: '30%', background: 'rgb(40, 40, 40)' }} key={key}>
      <audio src={url} style={{ width: '100%' }} controls />
      <div className={messageStyles.time}>{time}</div>
    </div>
  )
}

singleAudioMessage.propTypes = {
  url: PropTypes.string.isRequired,
  time: PropTypes.string.isRequired,
  whose: PropTypes.string.isRequired,
  key: PropTypes.number.isRequired,
}

async function getMedia() {
  let stream = null

  const constraints = { audio: true }
  stream = await navigator.mediaDevices.getUserMedia(constraints)

  return stream
}

class MessageForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      value: '',
      name: this.props.match.params.name,
      chatStyle: {},
      messages: [],
      chunks: [],
      audioIcon: voice,
      audioClick: null,
    }
    const allMessages = JSON.parse(localStorage.getItem(this.state.name)) || []
    for (let i = 0; i < allMessages.length; i += 1) {
      const currProps = {}
      currProps.time = allMessages[i].time
      currProps.whose = allMessages[i].whose
      currProps.key = i
      let currMessage = null
      if (allMessages[i].type === 'text') {
        currProps.content = allMessages[i].content
        currMessage = singleTextMessage(currProps)
      } else if (allMessages[i].type === 'image') {
        currProps.url = allMessages[i].url
        currMessage = singleImageMessage(currProps)
      } else if (allMessages[i].type === 'audio') {
        currProps.url = allMessages[i].url
        currMessage = singleAudioMessage(currProps)
      }
      this.state.messages.push(currMessage)
    }

    this.handleTextSubmit = this.handleTextSubmit.bind(this)
    this.handleTextChange = this.handleTextChange.bind(this)
    this.handleImageSubmit = this.handleImageSubmit.bind(this)
    this.handleImageChange = this.handleImageChange.bind(this)
    this.sendTextMessage = this.sendTextMessage.bind(this)
    this.sendImageMessage = this.sendImageMessage.bind(this)
    this.sendAudioMessage = this.sendAudioMessage.bind(this)
    this.sendGeo = this.sendGeo.bind(this)
    this.handleDragEnter = this.handleDragEnter.bind(this)
    this.handleDragLeave = this.handleDragLeave.bind(this)
    this.handleDragOver = this.handleDragOver.bind(this)
    this.handleDrop = this.handleDrop.bind(this)
    this.handleStartRecording = this.handleStartRecording.bind(this)
    this.handleEndRecording = this.handleEndRecording.bind(this)
    this.dataAvailable = this.dataAvailable.bind(this)
    this.stopRecording = this.stopRecording.bind(this)
    this.fileManager = React.createRef()
    this.messages = React.createRef()
    this.manageFiles = () => this.fileManager.current.click()

    this.mediaRecorder = null
  }

  // _____________geoposition_______________

  sendGeo(event) {
    event.preventDefault()
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition((position) => {
        const { latitude } = position.coords
        const { longitude } = position.coords
        const toSend = `https://www.openstreetmap.org/#map=18/${latitude}/${longitude}`
        this.sendTextMessage(toSend)
      })
    }
  }

  // _______________texts____________________

  sendTextMessage(content) {
    const currProps = {}
    currProps.content = content
    currProps.time = new Date().toLocaleTimeString(navigator.language, { hour: '2-digit', minute: '2-digit' })
    currProps.whose = 'mine'
    currProps.type = 'text'

    const allMessages = JSON.parse(localStorage.getItem(this.state.name)) || []
    currProps.key = allMessages.length
    allMessages.unshift(currProps)
    localStorage.removeItem(this.state.name)
    localStorage.setItem(this.state.name, JSON.stringify(allMessages))

    const currMessage = singleTextMessage(currProps)

    const currStateMess = this.state.messages
    currStateMess.unshift(currMessage)

    this.setState((state) => {
      return { messages: currStateMess }
    })
  }

  handleTextSubmit(event) {
    event.preventDefault()
    if (this.state.value !== '') {
      this.sendTextMessage(this.state.value)
      this.setState({ value: '' })
    }
  }

  handleTextChange(event) {
    event.preventDefault()
    this.setState({ value: event.target.value })
  }

  // _______________images__________________

  sendImageMessage(url, name) {
    const currProps = {}
    currProps.url = url
    currProps.time = new Date().toLocaleTimeString(navigator.language, { hour: '2-digit', minute: '2-digit' })
    currProps.whose = 'mine'
    currProps.type = 'image'
    currProps.name = name

    const allMessages = JSON.parse(localStorage.getItem(this.state.name)) || []
    currProps.key = allMessages.length
    allMessages.unshift(currProps)
    localStorage.removeItem(this.state.name)
    localStorage.setItem(this.state.name, JSON.stringify(allMessages))

    const currMessage = singleImageMessage(currProps)

    const currStateMess = this.state.messages
    currStateMess.unshift(currMessage)

    this.setState((state) => {
      return { messages: currStateMess, value: '' }
    })
  }

  handleImageSubmit(files) {
    if (files && files.length) {
      for (let i = 0; i < files.length; i += 1) {
        const src = window.URL.createObjectURL(files[i])
        const fileName = files[i].name

        const data = new FormData()
        data.append('image', files[i])
        fetch('https://tt-front.now.sh', {
          method: 'POST',
          body: data,
        }).then(() => {})

        this.sendImageMessage(src, fileName)
        this.messages.current.onload = () => {
          window.URL.revokeObjectURL(src)
        }
      }
    }
  }

  handleImageChange(event) {
    event.preventDefault()
    const upload = this.fileManager.current.files
    this.handleImageSubmit(upload)
  }

  // _______________audio____________________

  sendAudioMessage(url, name) {
    const currProps = {}
    currProps.url = url
    currProps.time = new Date().toLocaleTimeString(navigator.language, { hour: '2-digit', minute: '2-digit' })
    currProps.whose = 'mine'
    currProps.type = 'audio'
    currProps.name = name

    const allMessages = JSON.parse(localStorage.getItem(this.state.name)) || []
    currProps.key = allMessages.length
    allMessages.unshift(currProps)
    localStorage.removeItem(this.state.name)
    localStorage.setItem(this.state.name, JSON.stringify(allMessages))

    const currMessage = singleAudioMessage(currProps)

    const currStateMess = this.state.messages
    currStateMess.unshift(currMessage)

    this.setState((state) => {
      return { messages: currStateMess, value: '' }
    })
  }

  dataAvailable(event) {
    event.preventDefault()
    const toAssign = this.state.chunks
    toAssign.push(event.data)
    this.setState((state) => {
      return { chunks: toAssign }
    })
  }

  stopRecording(event) {
    this.setState({ audioIcon: voice, audioClick: this.handleStartRecording })

    const blob = new Blob(this.state.chunks, { type: 'audio/ogg; codecs=opus' })
    this.setState({ chunks: [] })

    const data = new FormData()
    data.append('audio', blob)
    fetch('https://tt-front.now.sh', {
      method: 'POST',
      body: data,
    }).then(() => {})

    const audioURL = URL.createObjectURL(blob)
    localStorage.setItem('audio', JSON.stringify(audioURL))

    this.sendAudioMessage(audioURL)
    this.messages.current.onload = () => {
      window.URL.revokeObjectURL(audioURL)
    }
  }

  async handleStartRecording(event) {
    event.preventDefault()
    this.setState({ audioIcon: stop, audioClick: this.handleEndRecording })

    const stream = await getMedia()
    this.mediaRecorder = new MediaRecorder(stream)
    this.setState({ chunks: [] })
    this.mediaRecorder.ondataavailable = this.dataAvailable
    this.mediaRecorder.onstop = this.stopRecording

    this.mediaRecorder.start()
  }

  handleEndRecording(event) {
    this.mediaRecorder.stop()
  }

  // _______________D&D______________________

  handleDragEnter(event) {
    event.preventDefault()
    this.setState({
      chatStyle: {
        border: '2px dashed rgb(255, 255, 255)',
      },
    })
  }

  handleDragLeave(event) {
    event.preventDefault()
    this.setState({ chatStyle: {} })
  }

  handleDragOver(event) {
    event.preventDefault()
  }

  handleDrop(event) {
    event.preventDefault()
    const upload = event.dataTransfer.files
    this.handleImageSubmit(upload)
    this.setState({ chatStyle: {} })
  }

  // ______________render____________________

  render() {
    return (
      <form className={formStyles.form} onSubmit={this.handleTextSubmit}>
        <Header leftImg={toChats} rightImg="" leftLink="/ChatList" name={this.state.name} onRightClick={() => {}} />
        <div
          className={formStyles.the_chat}
          style={this.state.chatStyle}
          onDragEnter={this.handleDragEnter}
          onDragLeave={this.handleDragLeave}
          onDragOver={this.handleDragOver}
          onDrop={this.handleDrop}
          ref={this.messages}
        >
          {this.state.messages}
        </div>
        <div className={formStyles.horizontal}>
          <input
            type="file"
            multiple
            accept="image/*"
            style={{ display: 'none' }}
            ref={this.fileManager}
            onChange={this.handleImageChange}
          />
          <img src={attach} alt="attachment" className={formStyles.img} onClick={this.manageFiles} />
          <img src={geo} alt="geolocation" className={formStyles.img} onClick={this.sendGeo} />
          <input
            type="text"
            value={this.state.value}
            placeholder="Your message"
            className={formStyles.input}
            onChange={this.handleTextChange}
          />
          <img
            src={this.state.audioIcon}
            alt="voice message"
            style={{ maxWidth: '29px' }}
            className={formStyles.img}
            onClick={this.state.audioClick || this.handleStartRecording}
          />
        </div>
        <div className={formStyles.empty} />
      </form>
    )
  }
}

MessageForm.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  match: PropTypes.object.isRequired,
}

export default MessageForm
