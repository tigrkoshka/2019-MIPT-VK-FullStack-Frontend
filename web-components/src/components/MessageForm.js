const template = document.createElement('template');
template.innerHTML = `
    <style>
        form {
            height: 100%;
            display: flex;
            flex-direction: column;
            overflow-x: hidden;
            transition: transform 2s;
            position: relative;
            left: -100%;
        }
        
        #header {
            display: flex;
            height: 7%;
        }
        
        .the-chat {
            overflow-y: scroll;
            display: flex;
            flex-direction: column-reverse;
            width: 100%;
            height: 100%;
        }
        
        .empty {
            flex-basis: 5px;
        }
        
        .curr-message {
            position: relative;
            max-width: 75%;
            margin-bottom: 10px;
            margin-top: 10px;
            border-radius: .8em;
        }
        
        .mine {
            margin-right: 5px;
            align-self: flex-end;
            background: #606163;
        }

        input[type=submit] {
            visibility: collapse;
        }  
    </style>
    <form class="form">
         <div class="the-chat"></div>
         <form-input name="message-text" placeholder="Введите сообщение"></form-input>
         <div class="empty"></div>
     </form>
`;

class MessageForm extends HTMLElement {
  constructor() {
    super();
    this._shadowRoot = this.attachShadow({ mode: 'open' });
    this._shadowRoot.appendChild(template.content.cloneNode(true));
    this.$form = this._shadowRoot.querySelector('.form');
    this.$input = this._shadowRoot.querySelector('form-input');
    this.$theChat = this._shadowRoot.querySelector('.the-chat');
    this.$messageFormHeader = this._shadowRoot.querySelector('message-form-header');

    this.$name = this.getAttribute('name');

    this.$messages = JSON.parse(localStorage.getItem(this.$name)) || [];

    this.$messageFormHeader = document.createElement('message-form-header');
    this.$messageFormHeader.id = "header";
    this.$messageFormHeader.name = this.$name;
    this.$form.insertBefore(this.$messageFormHeader, this.$form.firstChild);

    for (let i = 0; i < this.$messages.length; i += 1) {
      const currMessage = document.createElement('one-message');
      currMessage.content = this.$messages[i].content;
      currMessage.time = this.$messages[i].time;
      currMessage.classList.add('curr-message', 'mine');
      this.$theChat.insertBefore(currMessage, this.$theChat.firstChild);
    }

    this.$form.addEventListener('submit', this._onSubmit.bind(this));
    this.$form.addEventListener('keypress', this._onKeyPress.bind(this));
  }

  get form() {
    return this.$form;
  }

  get header() {
    return this.$messageFormHeader;
  }

  get name() {
    return this.$name;
  }

  _onSubmit(event) {
    event.preventDefault();
    const currText = this.$input.value.trim();
    if (currText.length === 0) {
      return;
    }

    const currMessage = document.createElement('one-message');
    currMessage.classList.add('curr-message', 'mine');
    currMessage.content = currText;
    currMessage.time = (new Date()).toLocaleTimeString(navigator.language, { hour: '2-digit', minute: '2-digit' });

    this.$messages.push({ content: currMessage.content, time: currMessage.time, name: currMessage.name });
    this.$theChat.insertBefore(currMessage, this.$theChat.firstChild);

    localStorage.removeItem(this.$name);
    localStorage.setItem(this.$name, JSON.stringify(this.$messages));

    this.$input.clear();
  }

  _onKeyPress(event) {
    if (event.keyCode === 13) {
      this.$form.dispatchEvent(new Event('submit'));
    }
  }
}

customElements.define('message-form', MessageForm);
