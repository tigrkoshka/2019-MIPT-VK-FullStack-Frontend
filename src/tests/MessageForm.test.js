/* eslint-disable react/jsx-props-no-spreading */
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import MessageForm from '../components/MessageForm'

describe('Chat page renders and acts correctly', () => {
  let props
  let component

  beforeEach(() => {
    props = {
      match: {
        params: {
          name: 'maslo',
          tag: '@maslo',
          userId: '1',
        },
      },
    }

    component = renderer
      .create(
        <MemoryRouter>
          <MessageForm {...props} />
        </MemoryRouter>,
      )
      .toJSON()
  })

  it('Should render normally', () => {
    expect(component).toMatchSnapshot()
  })
})
