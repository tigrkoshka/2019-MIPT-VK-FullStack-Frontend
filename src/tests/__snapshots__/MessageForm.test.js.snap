// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat page renders and acts correctly Should render normally 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="rectangle"
  >
    <div
      className="leftHorizontal"
    >
      <a
        className="leftButton"
        href="/ChatList/1"
        onClick={[Function]}
      >
        <div
          className="vertical"
        >
          <img
            alt=""
            height="25px"
            src="back.png"
            width="height"
          />
        </div>
        <div
          className="vertical"
        >
          <div
            className="text"
          >
            Chats
          </div>
        </div>
      </a>
    </div>
    <div
      className="horizontal"
    >
      <div
        className="vertical"
      >
        <div
          className="name"
        >
          maslo
        </div>
      </div>
    </div>
    <div
      className="rightHorizontal"
    />
  </div>
  <div
    className="the_chat"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    style={Object {}}
  />
  <div
    className="empty"
  />
</form>
`;
